command/unixOSCommand.go

```diff
@@ -40,19 +39,20 @@ func (unixOSCommand) CreateDirectory(
 
 func (unixOSCommand) CopyInlineFile(
 	runner *Runner,
-	name string,
 	fileContent pulumi.StringInput,
 	remotePath string,
 	useSudo bool,
-	append bool,
 	opts ...pulumi.ResourceOption) (*remote.Command, error) {
-	var createCmd pulumi.StringInput
-	if append {
-		createCmd = utils.AppendStringCommand(remotePath, useSudo)
-	} else {
-		createCmd = utils.WriteStringCommand(remotePath, useSudo)
+
+	sudo := ""
+	if useSudo {
+		sudo = "sudo"
 	}
-	return copyInlineFile(name, runner, fileContent, useSudo, createCmd, opts...)
+	backupPath := remotePath + "." + backupExtension
+	backupCmd := fmt.Sprintf("if [ -f '%v' ]; then %v mv -f '%v' '%v'; fi", remotePath, sudo, remotePath, backupPath)
```

Comment:
Adding this backup adds complexity but I don't see the added value given that we use automation.

Reply:
Isn't `deleteCmd` called even by automation, on stack update ?

Reply:
If you call `CopyInlineFile(content, remotePath1, false)` in an environment creation and then update the environment but without calling `CopyInlineFile(content, remotePath1, false)` then you want `remotePath1` to be removed.
For example, if you set the configuration of the Agent in one test but not in another one, then the change must be reverted.

Reply:
Yes but you don't a backup for that, you just need for `deleteCmd` to delete the file.

I think the thing you want to support is a file that was existing before any test (or at least before a `Stack/TestSuite` execution), is changed by `test1` then not changed by `test2` (so you'd like to revert to original file).
However I think this logic does not really work: if there's a change by `test1`, then a change by `test2`, then "not changed" by `test3`, you will rollback to `test2` version instead of rolling back to original version.

Also for `WithAgentConfig`, this is non-issue, the configuration must be given, in full, all the time. So not giving it implies empty `datadog.yaml` (which is default)

Reply:
> if there's a change by test1, then a change by test2, then "not changed" by test3, you will rollback to test2 version instead of rolling back to original version.

Consider the case with a file `foo` that contains `42`.
 - `test1` updates `foo` with `10` and create `foo.backup` with `42`
 - `test2` will update `foo`. The delete action of `test1` restore `42` in `foo`.  The `create` action create `foo.backup` with `42` and update `foo`
 - test3 call delete action of `test2` and restores `foo` with `42`.

Reply:
Yes okay with `DeleteBeforeReplace` it should work, however we're still dependent of whether the `command` provider considers that it should update the resource in-place or not. I'm guessing that an alteration of `createCommand` should trigger a new resource, but to be safe we should also set `ReplaceOnChanges`.

The code also needs to be commented because the Pulumi options are not in the same place as the code that requires it (which we should probably change as well).

Reply:
Code updated.